### PR TITLE
Mldb 1732 column count in where

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: sql/sql_expression.cc [hunk 1]
```diff
@@ -610,6 +610,10 @@ hasUnboundVariables() const
             return true;
     }
 
+    if (funcs.find("columnCount") != funcs.end()) {
```
Comment (steadtler) [inline sql/sql_expression.cc]: I understand how it fix your case, but thats not really the purpose of this method. If hasUnboundVariables is true, you would expect to have something in the variable or wildcards part of getUnbound.

**Comment by FinchPowers** [comment]
Ok je change ça. Et je viens de réaliser que le premier test échoue aussi. C'est un double bug fix.

File: core/dataset.cc [hunk 2]
```diff
@@ -1617,7 +1617,9 @@ generateRowsWhere(const SqlBindingScope & scope,
     UnboundEntities unbound = where.getUnbound();
 
     // Look for a free variable
-    bool needsColumns = unbound.hasUnboundVariables();
+    bool needsColumns = unbound.hasUnboundVariables() ||
+        unbound.funcs.find("columnCount") != unbound.funcs.end();
```
Comment (FinchPowers) [inline core/dataset.cc]: Corrige le problème du WHERE.

File: sql/sql_expression.cc [hunk 3]
```diff
@@ -3832,7 +3832,16 @@ SelectStatement::parse(ML::Parse_Context& context, bool acceptUtf8)
     skip_whitespace(context);
 
     //cerr << jsonEncode(statement) << endl;
-    
+
+    if (statement.from->getType() == "null") {
+        for (const auto & c: statement.select.clauses) {
+            if (typeid(*c.get()) == typeid(WildcardExpression)) {
+                throw HttpReturnException(
+                    400, "Wildcard usage requires a FROM statement");
```
Comment (FinchPowers) [inline sql/sql_expression.cc]: Corrige le problème de "SELECT *" sans FROM.

**Comment by FinchPowers** [comment]
On a un «problème» plus large où les fonctions ne sont pas reconnues s'il n'y a pas de FROM.
Exemple: `SELECT count(*)` retourne `Unable to find function 'count' binding function call 'count(*)'.  The function is not a built-in function, and either it's not a registered user function, or user functions are not available in the scope of the expression.`

File: sql/sql_expression.cc [hunk 4]
```diff
@@ -3832,7 +3832,16 @@ SelectStatement::parse(ML::Parse_Context& context, bool acceptUtf8)
     skip_whitespace(context);
 
     //cerr << jsonEncode(statement) << endl;
-    
+
+    if (statement.from->getType() == "null") {
+        for (const auto & c: statement.select.clauses) {
+            if (typeid(*c.get()) == typeid(WildcardExpression)) {
+                throw ML::Exception(
+                    "Wildcard usage requires a FROM statement");
```
Comment (FinchPowers) [inline sql/sql_expression.cc]: @jeremybarnes Ça ajoute des validations au parser. J'aimerais que tu donnes ton +1/-1 pour confirmer que ça va dans le sens que tu veux.

Comment (FinchPowers) [inline sql/sql_expression.cc]: @jeremybarnes ping

Comment (jeremybarnes) [inline sql/sql_expression.cc]: It doesn't go in the parser.  It should go in the query executor.  The parser deals with syntax, not semantics.

File: core/dataset.cc (hunk 2 above)
Comment (jeremybarnes) [inline core/dataset.cc]: This fix feels like a band-aid to me.  We should have an unbound.needsRow() function that looks for hasUnboundVariables() || hasRowFunction(), and implement those in some kind of a generic manner.

File: sql/sql_expression.cc [hunk 5]
```diff
@@ -620,6 +620,13 @@ hasUnboundVariables() const
     return false;
 }
 
+bool
+UnboundEntities::
+hasRowFunctions() const
+{
+    return funcs.find("columnCount") != funcs.end();
```
Comment (jeremybarnes) [inline sql/sql_expression.cc]: Add a comment here saying that we should make this more generic (nobody will remember to add other functions to this list).  Note that rowName(), rowHash(), rowPath(), leftRowHash(), rightRowHash(), etc are also functions that require the row.

**Comment by jeremybarnes** [comment]
OK to merge, but please add another issue to improve the implementation of the isRowFunction() function.  These should be marked by adding a field in BoundFunction describing if the function requires a row, and then collecting this information from the ExpressionValue info.

**Pull request merged**